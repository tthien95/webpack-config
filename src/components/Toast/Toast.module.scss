$noti-color: #fff;

@mixin hover-effect($opac) {
  opacity: $opac;
  cursor: pointer;
}

@mixin font-setting($font-size, $font-weight: 400) {
  font-size: $font-size;
  font-weight: $font-weight;
}

.notification-container {
  @include font-setting(14px);
  position: fixed;
  z-index: 999999;
  top: 12px;
  right: 12px;
  transition: transform 0.6s ease-in-out;
  animation: toast-in-right 0.7s;
}

.notification {
  background: $noti-color;
  transition: 0.3s ease;
  position: relative;
  pointer-events: auto;
  overflow: hidden;
  margin: 0 0 6px;
  padding: 30px;
  margin-bottom: 15px;
  width: 300px;
  max-height: 100px;
  border-radius: 3px 3px 3px 3px;
  box-shadow: 0 0 10px #999;
  color: #000;
  opacity: 0.9;
  background-position: 15px;
  background-repeat: no-repeat;

  &:hover {
    box-shadow: 0 0 12px $noti-color;
    @include hover-effect(1);
  }
}

.notification-title {
  @include font-setting(14px, 700);
  text-align: left;
  margin-top: 0;
  margin-bottom: 6px;
  width: 300px;
  height: 18px;
}

.notification-message {
  margin: 0;
  text-align: left;
  height: 18px;
  margin-left: -1px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.toast {
  height: 80px;
  width: 365px;
  color: $noti-color;
  padding: 15px;
}

.notification-container button {
  @include hover-effect(0.8);
  @include font-setting(16px, 700);
  position: relative;
  right: -0.3em;
  top: -0.3em;
  float: right;
  color: $noti-color;
  outline: none;
  border: none;
  text-shadow: 0 1px 0 $noti-color;
  line-height: 1;
  padding: 0;
  background: 0 0;
  border: 0;
}

@keyframes toast-in-right {
  from {
    transform: translateX(100%);
  }
  to {
    transform: translateX(0);
  }
}
