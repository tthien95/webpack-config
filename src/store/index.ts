import { configureStore } from '@reduxjs/toolkit';
import toastSlice from './toast-slice';

const store = configureStore({
  reducer: toastSlice,
});

export default store;
