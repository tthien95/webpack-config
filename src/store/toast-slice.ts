/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type Notification = {
  status: 'success' | 'error';
  title: string;
  message: string;
};

export interface ToastState {
  notification: Notification | null;
}

const initialState: ToastState = { notification: null };

const toastSlice = createSlice({
  name: 'toast',
  initialState,
  reducers: {
    showNotification: (state, action: PayloadAction<Notification>) => {
      const { status, title, message } = action.payload;
      state.notification = { status, title, message };
    },
    hideNotification: (state) => {
      state.notification = null;
    },
  },
});

export const toastActions = toastSlice.actions;

export default toastSlice.reducer;
