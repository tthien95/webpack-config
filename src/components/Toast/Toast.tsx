import { useEffect } from 'react';
import { createPortal } from 'react-dom';
import { useSelector, useDispatch } from 'react-redux';
import { toastActions, ToastState } from 'store/toast-slice';
import classes from './Toast.module.scss';

enum NotiStatus {
  success = '#5cb85c',
  error = '#d9534f',
}

const Toast = () => {
  const notification = useSelector((state: ToastState) => state.notification);
  const dispatch = useDispatch();

  useEffect(() => {
    if (notification) {
      const interval = setTimeout(() => {
        dispatch(toastActions.hideNotification());
      }, 3000);

      return () => {
        clearTimeout(interval);
      };
    }
    return undefined;
  }, [notification, dispatch]);

  if (!notification) {
    return null;
  }

  const { status, title, message } = notification;

  const content = (
    <div role="alert" className={classes['notification-container']}>
      <div
        className={`${classes.notification} ${classes.toast}`}
        style={{ backgroundColor: NotiStatus[status] }}
      >
        <button
          onClick={() => {
            dispatch(toastActions.hideNotification());
          }}
        >
          X
        </button>
        <div>
          <p className={classes['notification-title']}>{title}</p>
          <p className={classes['notification-message']}>{message}</p>
        </div>
      </div>
    </div>
  );

  return createPortal(content, document.getElementById('toastNoti')!);
};

export default Toast;
