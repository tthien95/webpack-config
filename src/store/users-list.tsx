/* eslint-disable @typescript-eslint/no-empty-function */
import React, { useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { AxiosError } from 'axios';
import { toastActions } from './toast-slice';
import { User, UsersListContextType } from '../type/user';

const UsersListContext = React.createContext<UsersListContextType>({
  usersList: [],
  setUsersList: () => {},
  isLoading: false,
  addUser: () => {},
  updateUser: () => {},
  deleteUser: () => {},
  setIsLoading: () => {},
  fnHandleError: () => {},
});

export const UserContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const [usersList, setUsersList] = useState<User[] | []>([]);
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useDispatch();

  const fnHandleError = useCallback(
    ({ response }: AxiosError<{ message: string }>) => {
      dispatch(
        toastActions.showNotification({
          status: 'error',
          title: 'Error',
          message:
            response?.data.message ||
            response?.statusText ||
            'There is something wrong happended while fetching data',
        })
      );
    },
    [dispatch]
  );

  const addUser = (value: User) => {
    const currUsersList = [...usersList];
    currUsersList.unshift(value);
    setUsersList(currUsersList);
  };

  const updateUser = (value: User) => {
    const currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == value.id);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1, value);
      setUsersList(currUsersList);
    }
  };

  const deleteUser = (userId: string) => {
    const currUsersList = [...usersList];
    // eslint-disable-next-line eqeqeq
    const userIndx = currUsersList.findIndex(({ id }) => id == userId);
    if (userIndx > -1) {
      currUsersList.splice(userIndx, 1);
      setUsersList(currUsersList);
    }
  };

  return (
    <UsersListContext.Provider
      value={{
        usersList,
        setUsersList,
        addUser,
        updateUser,
        deleteUser,
        isLoading,
        setIsLoading,
        fnHandleError,
      }}
    >
      {children}
    </UsersListContext.Provider>
  );
};

export default UsersListContext;
