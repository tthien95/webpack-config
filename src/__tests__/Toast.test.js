import { cleanup, render, waitFor, within } from '@testing-library/react';
import Toast from '../components/Toast/Toast';

import { useSelector, useDispatch, Provider } from 'react-redux';
import { createPortal } from 'react-dom';
import store from '../store/index';
import { toastActions } from '../store/toast-slice';
import userEvent from '@testing-library/user-event';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: jest.fn()
}));

const renderWithProvider = () => {
  return render(
    <Provider store={store}>
      <Toast />
    </Provider>
  );
};

describe('Navigation', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(() => {});
    createPortal.mockImplementation((component) => component);
  });
  afterAll(cleanup);

  it('should match snapshot for success', () => {
    useSelector.mockReturnValueOnce({
      status: 'success',
      title: 'Success',
      message: 'Success Message'
    });
    const { container } = renderWithProvider();
    expect(within(container).getByRole('alert')).toMatchSnapshot();
  });

  it('should match snapshot for error', () => {
    useSelector.mockReturnValueOnce({
      status: 'error',
      title: 'error',
      message: 'Error Message'
    });
    const { container } = renderWithProvider();
    expect(within(container).getByRole('alert')).toMatchSnapshot();
  });

  it('should not render toast when there is no notification', () => {
    useSelector.mockReturnValueOnce(null);
    const { container } = renderWithProvider();
    expect(within(container).queryByRole('alert')).not.toBeInTheDocument();
  });

  it('should trigger handler to close notification when press X button', () => {
    useSelector.mockReturnValueOnce({
      status: 'success',
      title: 'Success',
      message: 'Success Message'
    });

    const dispatch = jest.fn();

    useDispatch.mockReturnValueOnce(dispatch);

    const { container } = renderWithProvider();
    const noti = within(container).queryByRole('alert');
    const button = within(noti).getByRole('button');

    userEvent.click(button)

    expect(dispatch).toBeCalledTimes(1);
    expect(dispatch).toBeCalledWith(toastActions.hideNotification());
  });

  it('should trigger handler to close notification after 3 seconds', async () => {
    useSelector.mockReturnValueOnce({
      status: 'success',
      title: 'Success',
      message: 'Success Message'
    });

    const dispatch = jest.fn();

    useDispatch.mockReturnValueOnce(dispatch);

    renderWithProvider();

    await waitFor(() => expect(dispatch).toHaveBeenCalledTimes(1), {
      timeout: 3100
    });

    expect(dispatch).toBeCalledWith(toastActions.hideNotification());
  });
});
