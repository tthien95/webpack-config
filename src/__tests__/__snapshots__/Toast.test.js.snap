// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should match snapshot for error 1`] = `
<div
  class="notification-container"
  role="alert"
>
  <div
    class="notification toast"
    style="background-color: rgb(217, 83, 79);"
  >
    <button>
      X
    </button>
    <div>
      <p
        class="notification-title"
      >
        error
      </p>
      <p
        class="notification-message"
      >
        Error Message
      </p>
    </div>
  </div>
</div>
`;

exports[`Navigation should match snapshot for success 1`] = `
<div
  class="notification-container"
  role="alert"
>
  <div
    class="notification toast"
    style="background-color: rgb(92, 184, 92);"
  >
    <button>
      X
    </button>
    <div>
      <p
        class="notification-title"
      >
        Success
      </p>
      <p
        class="notification-message"
      >
        Success Message
      </p>
    </div>
  </div>
</div>
`;
